cip/CIP2015-09-16-public-type-system-type-annotation.adoc

```diff
@@ -0,0 +1,407 @@
+Note that this is primarily an exercise of consolidation rather than exploration of the broadest type system we could support.
+
+Furthermore, in order to support currently ongoing work for allowing users to define _stored procedures_, it has become necessary to specify types explicitly within signatures.
+This will help with the following:
+
+* In any interaction with an external language, type mapping between the external language and Cypher is necessary.
+Any such type mapping must rest on a well defined type system for Cypher in order to be consistent across different implementations.
+This applies to interaction via an external language supported by the implementation, as well as interaction via a remote protocol.
+* To call user defined procedures from within a larger Cypher query the planner needs to know static type signatures in order to generate efficient plans.
+* Statically typed signatures enable implementations to detect possible type errors at query compilation time.
+* Statically typed signatures also allow us to add support for procedure overloading in the future.
+
+To support these goals, it is necessary to at least define all valid, supported public types, introduce syntax for them as well as syntax for type annotations.
+
+=== Overview
+For the purposes of this CIP, a type system is a system of classifying expressions according to what kinds of values they may evaluate to during the execution of a Cypher query. Furthermore, for a given type system, a type annotation enables annotating terms with type expectations for later re-use in determining the type of other expressions (e.g. when determining the type of a function call expression and its arguments based on type annotations used in function signatures). The type system presented in this CIP does not include a specification of a type inferencing and checking algorithm for Cypher queries (out of scope).
+
+Cypher today has a rich type system that unifies Java implementation types into a coherent type hierarchy and an accompanying type checker that supports rich types with https://en.wikipedia.org/wiki/TypeParameter[type parameters] and https://en.wikipedia.org/wiki/Union_type[union types].
+So far it has not been possible to explicitly specify the type of an expression in a query.
+
+This CIP proposes to introduce a well defined type system, syntax for type literals and for type annotations for public types.
+Any implementation may choose to incorporate a set of internal types beyond those specified here.
+All such internal types must map cleanly onto the public type system and should generally not be made visible to end users.
+
+The details of internal types are out of scope for the purposes of this document.
+
+== Proposal
+
+=== Types and Type Literal Syntax
+
+The section below presents presents the new Public Type System (PTS) proposed for Cypher.
+
+==== Material and nullable types
```

Comment:
It would be helpful to discuss _why there is a need for (non)nullable types_. Requiring non-null values is typically a runtime concern in a language. The current Neo4j implementation of cypher treats all types as nullable, and warning/error messages containing type names do not distinguish between them. If this CIP was to be accepted, an error message containing "String" would more correctly, yet oddly, contain "String?".